cleanup of work queue API documentation and minor behavior changes

pabigot opened the pull request:
This is an alternative to #28579 that does not change the existing behavior related to cancelling delayed work items, but does update the documentation to describe that behavior as accurately as possible, including more about its limitations.

It also fixes a minor race condition and extends the tests to improve coverage and verify the return values in cases where the states can be observed.  This increases coverage of kernel/work_q to 100%.

pabigot committed:
kernel: delayed_work: fix race condition in k_delayed_work_cancel

The implementation checks the work queue pointer before taking a lock
and entering code that, depending on build options, may fault if that
pointer is null.  Move the check under the lock to ensure condition is
preserved.  Then the check in work_cancel is no longer necessary since
condition is verified under lock at all call sites.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot committed:
kernel: delayed_work: note restriction in re-using delayed work items

The delayed work work_q pointer is set on submission, and remains set
even after the work item completes.  Attempts to re-submit the
completed work item to a different queue will be rejected because the
code cannot distinguish a completed item from one that is still
waiting for a delay or to be processed.

Document this restriction.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot committed:
kernel: delayed_work: update k_delayed_work_cancel documentation

Although the documentation states that only work items that have not
completed the delay can be cancelled, in fact pending items can also
be removed from a work queue.  Document that behavior

Also document the specific return value that will be obtained based on
the state of the work item at the time cancellation was attempted
using the current implementation.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot committed:
doc: kernel: workqueue: correct documentation of delayed work API

The delayed work API has been changed to allow cancellation in
conditions not previously documented, but this feature can cause both
submission and cancellation to fail in the general case.  Summarize
the conditions and highlight the importance of checking result codes
for these functions.

Also explicitly note that no kernel API can reliably indicate that a
work item has been completed, and that doing so is the responsibility
of the code that provides the work handler.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot committed:
tests: kernel: work_queue_api: clean up work_pending arguments

Pass a pointer to the work item member rather than casting the
augmented work item pointer to a base work item pointer.

Also the return type of k_work_pending() is bool, so use that rather
than comparing it to zero.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot committed:
tests: kernel: work_queue_api: coverage for delayed pending

This API function wasn't being tested.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot committed:
tests: kernel: work_queue_api: validate legacy multiq behavior

The current implementation of delayed work retains a pointer to the
queue unless the work item is successfully cancelled, preventing a
completed item from being resubmitted to a different queue.  Confirm
this behavior and its workaround.

Also validates some unsuccessful cancel return values.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot committed:
tests: kernel: work_queue_api: validate legacy no-wait behavior

The current implementation of delayed work will cancel and re-submit a
pending work item that is no-wait, putting it at the back of the
queue.  Verify this behavior.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

Vudentz reviewed (approved):
LGTM

nashif commented on kernel/work_q.c:
```diff
@@ -132,12 +128,12 @@ int k_delayed_work_submit_to_queue(struct k_work_q *work_q,
-	int ret = work_cancel(work);
+	int ret = -EINVAL;
+
+	if (work->work_q) {
```
You need to explicitly compare pointer with NULL (misra 14.4)

pabigot replied:
done

pabigot replied:
@nashif please revisit, thanks.

pabigot committed:
tests: kernel: work_queue_api: align schedule tests to tick

Detection of transition from delayed to pending can fail in some cases
if the timeouts are not precisely managed.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot commented:
Added timer alignments to the delayed work tests to avoid  intermittent failures on mps2_an521.

pabigot committed:
kernel: work: clarify effect of resubmitting pending items

Make more clear that submitting a work item to one queue while it is
pending on another queue has no effect.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>
